src/tls/certificate.rs

```diff
@@ -47,28 +47,19 @@ pub(crate) fn make_cert(keypair: &Keypair) -> Result<rcgen::Certificate, super::
         let mut buf =
             [0u8; LIBP2P_SIGNING_PREFIX_LENGTH + LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH];
         buf[..LIBP2P_SIGNING_PREFIX_LENGTH].copy_from_slice(&super::LIBP2P_SIGNING_PREFIX[..]);
-        buf[LIBP2P_SIGNING_PREFIX_LENGTH..].copy_from_slice(certif_pubkey);
+        buf[LIBP2P_SIGNING_PREFIX_LENGTH..].copy_from_slice(&certif_pubkey);
         keypair.sign(&buf)?
     };
 
     // Generate the libp2p-specific extension.
     let libp2p_extension: rcgen::CustomExtension = {
         let extension_content = {
             let serialized_pubkey = keypair.public().into_protobuf_encoding();
-            yasna::construct_der(|writer| {
-                writer.write_sequence(|writer| {
-                    writer
-                        .next()
-                        .write_bitvec_bytes(&serialized_pubkey, serialized_pubkey.len() * 8);
-                    writer
-                        .next()
-                        .write_bitvec_bytes(&libp2p_ext_signature, libp2p_ext_signature.len() * 8);
-                })
-            })
+            yasna::encode_der(&(serialized_pubkey, libp2p_ext_signature))
         };
 
         let mut ext = rcgen::CustomExtension::from_oid_content(LIBP2P_OID, extension_content);
-        ext.set_criticality(true);
+        ext.set_criticality(false);
```

Comment:
From an X.509 perspective, the extension is critical, in that the whole certificate is useless to a peer that cannot interpret it.  That said, if using `false` here improved interoperability, I would be okay with it.

Reply:
According to https://github.com/libp2p/specs/blob/master/tls/tls.md:

> This extension MAY be marked critical.

I tried keep it, but the go implementation fails to connect:

```
go-libp2p-quic-transport/cmd$ go run client/main.go /ip4/127.0.0.1/udp/8383/quic 12D3KooWKnmstT37WmaiMRviHevoHZvWBrL7KsAXddcgrzqpp79i
2021/07/30 15:02:04 Dialing /ip4/127.0.0.1/udp/8383/quic
2021/07/30 15:02:04 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for details.
2021/07/30 15:02:04 CRYPTO_ERROR (0x12a): certificate verification failed: x509: unhandled critical extension
exit status 1
```

Reply:
That's right, here is the bug: https://github.com/libp2p/go-libp2p-tls/blob/master/crypto.go#L116.

However the extension MAY be marked as critical, so we don't have to.

Reply:
so unresolving this because it seems it's go-libp2p that is broken